Assets/_Dev/Runtime/Spawner.cs

```diff
@@ -8,62 +8,123 @@ namespace Playground
     public class Spawner : MonoBehaviour
     {
         [Header("Wave Definition")]
-        [SerializeField, Tooltip("The enemy prefabs to spawn.")]
-        private BasicEnemyController[] enemyPrefabs;
-        [SerializeField, Tooltip("The number of enemies to spawn to spawn each second.")]
-        private float spawnRate = 1f;
+        [SerializeField, Tooltip("The list of defined waves used for spawning.")]
+        private WaveDefinition[] waves;
         [SerializeField, Tooltip("The radius around the spawner to spawn enemies.")]
         internal float spawnRadius = 5f;
-        [SerializeField, Tooltip("The duration of each spawn wave in seconds.")]
-        private float waveDuration = 10f;
         [SerializeField, Tooltip("The duration of the wait between each spawn wave in seconds.")]
         private float waveWait = 5f;
+        [SerializeField, Tooltip("When defined waves are exhausted, generate more?.")]
+        private bool generateWaves = true;
 
         [Header("Events")]
         [SerializeField, Tooltip("The event to trigger when this spawner is destroyed.")]
         public UnityEvent onDestroyed;
+        [SerializeField, Tooltip("The event to trigger when all waves are complete.")]
+        public UnityEvent onWavesComplete;
```

Comment:
I missed the 's' here and thus read it as `OnWaveComplete`. Should it be `OnAllWavesComplete`?

Perhaps more importantly, this will not be called if `generateWaves` is `true` (the default). Should we allow a limit to the number of waves? Not a blocker on merging, just thinking in the review.

Reply:
good question, I guess they could always be infinite waves, I was just thinking that maybe some of them we would want the option to be finite. maybe the spawners have resources to "power" the spawning; or maybe when they run out of waves they turn into enemies themselves? just thinking (feature creep).

Reply:
I think the option is good. I'm OK with merging like this, but perhaps we should have a `numberOfWaves` where 0 is infinite.